fix(TimeDistributed): error with custom layer output_shape tuples

justindujardin opened the pull request:
**Problem**
When using custom layers with the `TimeDistributed` wrapper in the v2 preview, an error is thrown if the output_shape is not returned as a `TensorShape` instance. This goes contrary to the v1 docs, and also how custom layers work without the time distributed wrapper. 

**Reproduction**
https://colab.research.google.com/gist/justindujardin/d22de372c3eb6adaf7b6d85c8f6c5dfb/tf_keras_layer_output_shape_bug.ipynb

**Solution**
Coerce tuple/list output shape types into TensorShape without erroring.

justindujardin committed:
fix(TimeDistributed): do not throw when tuple or list is returned from compute_output_shape

 - the docs show custom layers returning a tuple or list to describe output_shape. in v2 TimeDistributed custom layers fail if the output_shape is not a TensorShape instance.
 - instead of failing, coerce to TensorShape if one is not returned.

justindujardin committed:
test(TimeDistributed): assert supported output_shape types

 - list/tuple/TensorShape all end up as TensorShape in the end

qlzh727 commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -290,6 +291,43 @@ def call(self, inputs):
+                   TensorShapeOutputShapeLayer]
+    # Custom layers can specify output shape as a list, tuple, or TensorShape
+    for layer in test_layers:
+      td = layer()
```
What is td? Can we have a more readable variable name here?

justindujardin replied:
I originally missed the TimeDistributed wrapper, thanks for catching it.

qlzh727 commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -290,6 +291,43 @@ def call(self, inputs):
+    # Custom layers can specify output shape as a list, tuple, or TensorShape
+    for layer in test_layers:
+      td = layer()
+      ph = keras.backend.placeholder(shape=(None, 1, 13))
```
I guess this is a input tensor, let's call it "inputs"

justindujardin replied:
I removed this because I was only testing the output_shape, not the actual layer calls.

qlzh727 commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -290,6 +291,43 @@ def call(self, inputs):
+      ph = keras.backend.placeholder(shape=(None, 1, 13))
+      out = td(ph)
+      # all output shapes are exposed in the end as TensorShape instances
+      self.assertEqual(isinstance(out, tensor_shape.TensorShape), True)
```
please use self.assertIsInstance()

justindujardin replied:
Done

qlzh727 commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -290,6 +291,43 @@ def call(self, inputs):
+      ph = keras.backend.placeholder(shape=(None, 1, 13))
+      out = td(ph)
+      # all output shapes are exposed in the end as TensorShape instances
+      self.assertEqual(isinstance(out, tensor_shape.TensorShape), True)
```
Since the out here is an output tensor, I don't feel the assertion here is correct. Shouldn't u use the time distributed wrapper to wrap those test layer and assert the output shape? Checking the type here is probably not enough.

Also if you want check the type, probably should use compute_output_shape explicitly.

justindujardin replied:
I admit the test is speculative based on the surrounding tests. I haven't been able to run it locally yet because my local bazel build failed. I will execute the test and verify that it conforms to your suggested changes.

qlzh727 replied:
This is still not correct, the output is a time distributed wrapped input, which is not a tensor shape.
You should assert the shape of the output, instead of the type of the output.

justindujardin replied:
Thanks for the review, and sorry for the confusion; I pushed the first commit eagerly before I read all of your comments. I've now incorporated your feedback and verified the test passes locally with Bazel.

qlzh727 reviewed (changes requested):
Thanks for the fix. The fix looks good, just some question about the tests.

justindujardin committed:
chore: fix missing TimeDistributed layer in test

 - cleanup variable names from review

justindujardin committed:
chore: simplify test based on feedback

 - verify it passes running locally with bazel
 - remove call methods since we're just measuring the output_shape and making sure the wrapper in TimeDistributed doesn't crash like it did before.

justindujardin committed:
fix(BiDirectional): coerce return type of compute_output_shape

 - to TensorShape if it's specified as a tuple or list.
 - add test to verify that wrapper no longer throws

justindujardin commented:
While writing the test I noticed the same error was in the `BiDirectional` class wrapper so I added a similar fix and test for it.

qlzh727 commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -715,6 +708,36 @@ def test_Bidirectional_with_constants_layer_passing_initial_state(self):
+    for layer in test_layers:
+      input_layer = keras.layers.Bidirectional(layer(3))
+      input_shape = (None, 1, 13)
+      output_shape = input_layer.compute_output_shape(input_shape)
```
Same for this test case, I think the test layer will still have the same issue as the output's shape from call() is different from compute_output_shape().

justindujardin committed:
chore: update tests to call test layers

 - as per review, make the tests more general than a specific regression on mishandling the result of `compute_output_shape`

justindujardin commented on tensorflow/python/keras/layers/wrappers_test.py:
```diff
@@ -323,10 +314,12 @@ def compute_output_shape(self, input_shape):
-      placeholder = keras.backend.placeholder(shape=(None, 1, 13))
-      out = time_distributed(placeholder)
+      input_shape = (None, 1, 13)
+      output_shape = time_distributed.compute_output_shape(input_shape)
```
I'm not sure I agree that the test itself is bad. My change introduces a fix for improper handling of the return value of `compute_output_shape` in the TimeDistributed and BiDirectional wrapper classes. By invoking that method I have constructed tests that fail prior to my fix and pass after it is applied.

That said, I updated the tests to make the changes you asked for.
